pkg/store/bucket.go

```diff
@@ -1272,25 +1332,55 @@ func (s *BucketStore) LabelValues(ctx context.Context, req *storepb.LabelValuesR
 		resHints.AddQueriedBlock(b.meta.ULID)
 
 		indexr := b.indexReader(gctx)
-		extLabels := b.meta.Thanos.Labels
 
 		g.Go(func() error {
 			defer runutil.CloseWithLogOnErr(s.logger, indexr, "label values")
 
-			// Do it via index reader to have pending reader registered correctly.
-			res, err := indexr.block.indexHeaderReader.LabelValues(req.Label)
-			if err != nil {
-				return errors.Wrap(err, "index header label values")
-			}
+			var result []string
+			if len(reqSeriesMatchers) == 0 {
+				// Do it via index reader to have pending reader registered correctly.
+				res, err := indexr.block.indexHeaderReader.LabelValues(req.Label)
+				if err != nil {
+					return errors.Wrapf(err, "index header label values for block %s", b.meta.ULID)
+				}
+
+				// Add the external label value as well.
+				if extLabelValue := b.extLset.Get(req.Label); extLabelValue != "" {
+					res = strutil.MergeSlices(res, []string{extLabelValue})
+				}
+				result = res
+			} else {
+				seriesSet, _, err := blockSeries(b.extLset, indexr, nil, reqSeriesMatchers, nil, seriesLimiter, true, req.Start, req.End, nil)
```

Comment:
This can be optimized with https://github.com/prometheus/prometheus/pull/8400.
We can open a new issue to track it and see if someone would like to contribute to this.

Reply:
**This** PR already uses the same technique, i.e. finding postings for matching series, and only extracting label values from them (see `labelValuesWithMatchers` in https://github.com/prometheus/prometheus/pull/8400). We cannot use https://github.com/prometheus/prometheus/pull/8400 directly, because we don't have all required information (ie. postings) locally in the index header. That's why we use `blockSeries`, which fetches postings for us.

Reply:
That makes sense. Thanks for the explanation!